### DISCUSSION TITLE
Why can't we use Google Cloud KMS for Transaction Signers?

### DISCUSSION THREAD

**Discussion by yourbuddyconner** [celo-org/optics-monorepo#598] [Q&A]
All of our infrastructure is in Google Cloud presently, why can't we use the native HSM service that Google Offers?

**Comment by yourbuddyconner** [comment]
We'd love to use [Google KMS](https://cloud.google.com/security-key-management) but their HSM implementation doesn't support `Secp256k1` with `keccak256` or a raw digest signing mode.

**Reply by prestwich** [reply]
to expand:

- ECDSA is not secure unless both signer and verifier independently produce the digest using a cryptographic hash function
- To prevent misuse, it is common for cryptographic APIs to forbid the API user from signing a digest unless the API  produces the digest internally
- GCP supports only `sha256`, and has no raw digest-signing mode
- Ethereum requires `keccak256`
- AWS exposes a raw digest signing mode
- We use the raw digest signing mode + local production of the digest to sign ethereum txns and messages

**Reply by bdhess** [reply]
I'm not sure what's meant by "raw digest signing mode".  Does this mean signing over a precomputed digest?  If so, I think Google Cloud KMS contains everything you need, though this isn't totally obvious.

The KMS [algorithm](https://cloud.google.com/kms/docs/algorithms#elliptic_curve_signing_algorithms) `EC_SIGN_SECP256K1_SHA256` is meant to be used with a precomputed SHA-256 digest.  However, the ECDSA signature algorithm doesn't encode any information about the digest into the signature, and Cloud KMS has no way of knowing whether a supplied 32-byte digest is the output from SHA-256, or from some other 256-bit hash function.  As such, a keccak256 digest could be supplied in the `sha256` request field, and a valid ECDSA+keccak256 signature would be emitted by Cloud KMS.

This is sort of a gap/unintended consequence of KMS' design, but it absolutely works.  Even if doing something to "fix" this were possible (it isn't), such a change would be breaking.  All this is to say that what I'm proposing is safe for use in production.

If you'd like for this to be confirmed with GCP Support, please feel free to raise a support ticket that references this issue.

**Reply by guenoledc** [reply]
Hi @bdhess Have you been able to make any progress on this since this ticket ?

I am also trying to implement the eth signer with GCP and found the same hitch with keccak256. 
I tried to pass it to the `asymmetricSign` function (I use javascript) and received some signature.

Now I have an issue : 
- the returned signature is not 64 bytes like for a normal signature (70 or 71 bytes long, the initial part seems to be a prefix!)
- the public key of the `EC_SIGN_SECP256K1_SHA256` generated key does not seems to be 64 bytes long either (more 88)

Did you figure something out ?

**Reply by guenoledc** [reply]
Hi all

I managed to make it working quite fine with keccak256 of the hash and derive the recovery id.
Took me a while so it's worth sharing hoping it will help others and eventually get some feedbacks

Here are some code extract (based on the code sample)

```js
async function signEthereum(keyPath, message) {
  const crypto = require('crypto');

  const pubKey = await client.getPublicKey({name: keyPath})
  const p2 = crypto.createPublicKey(pubKey[0].pem)
  let pubKeyBuf = p2.export({format:"der", type:"spki"});
  pubKeyBuf = pubKeyBuf.slice(pubKeyBuf.length-65) // uncompressed public key starting with 0x04
  
  // calculate the hash with the keccak256 function 
  const {keccak256} = require("ethereum-cryptography/keccak")
  const digest =keccak256(Uint8Array.from(message))

  // Optional but recommended: Compute digest's CRC32C.
  // Ensure fast-crc32c has been installed, `npm i fast-crc32c`.
  const crc32c = require('fast-crc32c');
  const digestCrc32c = crc32c.calculate(digest);

  // Sign the message with Cloud KMS
  const [signResponse] = await client.asymmetricSign({
    name: keyPath,
    digest: {
      sha256: digest
    },
    digestCrc32c: {
      value: digestCrc32c,
    },
  });

  // Optional, but recommended: perform integrity verification on signResponse.
  // For more details on ensuring E2E in-transit integrity to and from Cloud KMS visit:
  // https://cloud.google.com/kms/docs/data-integrity-guidelines
  if (signResponse.name !== keyPath) {
    throw new Error('AsymmetricSign: request corrupted in-transit');
  }
  if (!signResponse.verifiedDigestCrc32c) {
    throw new Error('AsymmetricSign: request corrupted in-transit');
  }
  if (
    crc32c.calculate(signResponse.signature) !==
    Number(signResponse.signatureCrc32c.value)
  ) {
    throw new Error('AsymmetricSign: response corrupted in-transit');
  }
  const { signatureImport, signatureNormalize } = require("secp256k1");
  // Very important, lost a lot of time on it, convert the returned signature into a 64 bytes long valid signature
  let _64 = signatureImport(signature);
  // this is to get the low part of the curve
  const normalized = signatureNormalize(_64); 
  const r = normalized.slice(0, 32)
  const s = normalized.slice(32, 64)
  // this is to find the appropriate recovery id (below)
  const recId = calculateRecoveryId(normalized, digest, pubKeyBuf); 
  // to embed into the recovery id the chain ID for EIP-155
  const v = calculateV(1 /* ethereum chainId */, recId); 

  return {r, s, v}
}

function calculateRecoveryId(signature, hash, uncompressPubKey) {
  const { ecdsaRecover } = require("secp256k1");
  let recId = -1;

  for (let i = 0; i < 4; i++) {
    // try with a recoveryId of i
    const rec = ecdsaRecover(signature, i, hash, false);
    if (Buffer.compare(rec, uncompressPubKey) == 0) {
      recId = i;
      break;
    }
  }
  if (recId == -1)
    throw new Error("Impossible to calculare the recovery id. should not happen");
  return recId;
}

function calculateV(chainId, recovery) {
    if (!chainId || typeof chainId === 'number') {
      // return legacy type ECDSASignature (deprecated in favor of ECDSASignatureBuffer to handle large chainIds)
      if (chainId && !Number.isSafeInteger(chainId)) {
        throw new Error('The provided number is greater than MAX_SAFE_INTEGER (please use an alternative input type)');
      }

      const v = chainId ? recovery + (chainId * 2 + 35) : recovery + 27;
      return v;
    } else {
      throw new Error("Other chainId type not implemented");
    }
  }

```

**Reply by 3266miles** [reply]
Hi @guenoledc this is really cool. Keen to take it for a test this weekend myself. How confident are you in its working given that its sha256 and not keccak256 in the signature? Surely that will produce issues in other places?

**Reply by guenoledc** [reply]
@3266miles

I cannot yet tell if there are impacts down the line but you can pass whatever 32bytes ling buffer as hash and KMS won't care. 

Then down the line I have submitted signed transactions to ethereum (test net) and it works so i suppose it will work all the time. 

Need to battle test it a bit more with a long sequence of transactions 

Let me know if you find some issues pls

**Reply by EWCunha** [reply]
Hi @guenoledc. I'm also trying to use GCP KMS to sign Ethereum transactions using Python, but weirdly I'm not getting the same signature when a provide the exact same transaction. Does this happen to you as well?

**Reply by guenoledc** [reply]
> Hi @guenoledc. I'm also trying to use GCP KMS to sign Ethereum transactions using Python, but weirdly I'm not getting the same signature when a provide the exact same transaction. Does this happen to you as well?

@EWCunha 

No I did bot experience this but did bot try on python 

Can I recommend that you do a step by step comparison (maybe using a debugging session) and find which step is starting to differs. Often what we see is different parameters interpretation in the inner libraries of the languages. 

Good luck and share your findings to other

**Reply by bdhess** [reply]
GCP KMS signatures are not deterministic.  It is expected that the signature differs with repeated signature calls over the same data.

**Reply by guenoledc** [reply]
So I guess, the only thing to do to test is try to decode the signature with a [ecrecover](https://docs.web3js.org/api/web3-eth-accounts/function/ecrecover/) call. If it returns the public key of the private key you are trying to sign with then the signature (r, s, v) is valid from an eth perspective.
